stdlib/public/core/Duration.swift

```diff
@@ -0,0 +1,205 @@
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+@available(SwiftStdlib 9999, *)
+@frozen
+public struct Duration: Sendable {
+  /// The low 64 bits of a 128-bit signed integer value counting attoseconds.
+  @usableFromInline
+  internal var _low: UInt64
+
+  /// The high 64 bits of a 128-bit signed integer value counting attoseconds.
+  @usableFromInline
+  internal var _high: Int64
+
+  internal init(_attoseconds: _Int128) {
+    self._low = _attoseconds.low
+    self._high = _attoseconds.high
+  }
+
+  internal init(_seconds: Int64, nanoseconds: Int64) {
+    self = Duration.seconds(_seconds) + Duration.nanoseconds(nanoseconds)
+  }
+
+  internal var _attoseconds: _Int128 {
+    _Int128(high: _high, low: _low)
+  }
+
+  public var seconds: Int64 {
+    Int64(_attoseconds / 1_000_000_000_000_000_000)
+  }
+
+  public var nanoseconds: Int64 {
+    let seconds = _attoseconds / 1_000_000_000_000_000_000
+    let nanoseconds =
+      Int64((_attoseconds - seconds * 1_000_000_000_000_000_000) / 1_000_000_000)
+    return nanoseconds
+  }
+}
+
+@available(SwiftStdlib 9999, *)
+extension Duration {
+  @available(SwiftStdlib 9999, *)
+  public static func seconds<T: BinaryInteger>(_ seconds: T) -> Duration {
+    return Duration(_attoseconds: _Int128(seconds) *
+                                 1_000_000_000_000_000_000)
+  }
+
+  @available(SwiftStdlib 9999, *)
+  public static func seconds(_ seconds: Double) -> Duration {
+    return Duration(_attoseconds: _Int128(seconds *
+                                 1_000_000_000_000_000_000))
+  }
+
+  @available(SwiftStdlib 9999, *)
+  public static func milliseconds<T: BinaryInteger>(
+    _ milliseconds: T
+  ) -> Duration {
+    return Duration(_attoseconds: _Int128(milliseconds) *
+                                 1_000_000_000_000_000)
+  }
+
+  @available(SwiftStdlib 9999, *)
+  public static func milliseconds(_ milliseconds: Double) -> Duration {
+    return Duration(_attoseconds: _Int128(milliseconds *
+                                 1_000_000_000_000_000))
+  }
+
+  @available(SwiftStdlib 9999, *)
+  public static func microseconds<T: BinaryInteger>(
+    _ microseconds: T
+  ) -> Duration {
+    return Duration(_attoseconds: _Int128(microseconds) *
+                                 1_000_000_000_000)
+  }
+
+  @available(SwiftStdlib 9999, *)
+  public static func microseconds(_ microseconds: Double) -> Duration {
+    return Duration(_attoseconds: _Int128(microseconds *
+                                 1_000_000_000_000))
+  }
+
+  @available(SwiftStdlib 9999, *)
+  public static func nanoseconds<T: BinaryInteger>(
+    _ nanoseconds: T
+  ) -> Duration {
+    return Duration(_attoseconds: _Int128(nanoseconds) *
+                                 1_000_000_000)
+  }
+}
+
+@available(SwiftStdlib 9999, *)
+extension Duration: Codable { }
```

Comment:
Is the plan for `Duration` to use synthesized `Codable` conformance, or is a custom implementation in the works?

I think the synthesized conformance would encode the `_low` and `_high` values directly. For durations larger than ~10ms (?), the `_low` value will be greater than the upper bound for full-precision `Number`s in JavaScript (i.e. 2^53).

It seems like it might be prudent to avoid exceeding that limit. (Not only to avoid loss of precision, but also to avoid parsing errors due to the number being "too big".)

Encoding the "components" (full-seconds, and subsecond-attoseconds) might be safer. Or encoding the entire 128-bit number as a string.

Reply:
I've been thinking about this specifically with regards to the components discussion. I think you are right; string encoding may be slow but it won't impact accuracy. Having a round trip of all valid values seems like the right thing to do. I will have some modifications to this on Monday.